classic_cp/kidneyexchange/kidneyexchange.jl

```diff
@@ -8,98 +8,124 @@ end
+    # xEqualIndex_Bool[i] = true => i ∉ cycle
+    # xEqualIndex_Bool[i] = false => i ∈ cycle
+    xEqualIndex_Bool = Vector{SeaPearl.AbstractBoolVar}(undef, n)  
+
+    # "view" variable to be able to use the "ReifiedInSet" constraint
+    # xNotEqualIndex_Int[i] = 1 => i ∉ cycle
+    # xNotEqualIndex_Int[i] = 0 => i ∈ cycle
+    xNotEqualIndex_Int = Vector{SeaPearl.AbstractIntVar}(undef, n) 
+
+    # "fake" variable to know if xNotEqualIndex_Int == 0
+    zeroSet = SeaPearl.IntSetVar(0, 0, "zeroSet", model.trailer)
+    # require the unique value to ensure that zeroSet is bounded
+    SeaPearl.require!(zeroSet.domain, 0)
 
     for i = 1:n
         x[i] = SeaPearl.IntVar(1, n, "x_"*string(i), model.trailer)
         SeaPearl.addVariable!(model, x[i]; branchable=true)
-        x_offset[i] = SeaPearl.IntVarViewOffset(x[i], -i, x[i].id*"-"*string(i))
+
+        index[i] = SeaPearl.IntSetVar(i, i, "index_"*string(i), model.trailer)
+        # require the unique value to ensure that isbound(index[i]) return true
+        SeaPearl.require!(index[i].domain, i)
+        
+        xEqualIndex_Bool[i] = SeaPearl.BoolVar("xEqualIndex_Bool_"*string(i), model.trailer)
+        xNotEqualIndex_Int[i] = SeaPearl.IntVar(0, 1,"xNotEqualIndex_Int_"*string(i), model.trailer)
     end
 
     ### Constraints ###
     for i = 1:n
+        #Fix xEqualIndex_Bool
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(x[i], index[i], xEqualIndex_Bool[i], trailer))
+
+        #Fix xNotEqualIndex_Int
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(xNotEqualIndex_Int[i], zeroSet, xEqualIndex_Bool[i], trailer))
+
+        #Add incompatibilities to the model
         for j in 1:n
-            if j ∉ c[i]
+            if j ∉ c[i] && i !=j
                 SeaPearl.addConstraint!(model, SeaPearl.NotEqualConstant(x[i], j, trailer))
             end
         end
     end
-    SeaPearl.addConstraint!(model, SeaPearl.AllDifferent(x, trailer))
 
-    # additional constraint: sum(x_offset) = 0
+    #Check that any pair receives/gives more than 1 kidney and that for each pair "give a kidney <=> receive a kidney"
+    SeaPearl.addConstraint!(model, SeaPearl.AllDifferent(x, trailer))
 
     ### Objective ###
-    numberOfExchanges = SeaPearl.IntVar(-n, 0, "numberOfExchanges", trailer) 
-    SeaPearl.addVariable!(model, numberOfExchanges; branchable=true)
+
+    #SeaPearl's solver minimize the objective variable, so we use minusNumberOfExchanges in order to maximize the number of exchanges
+    minusNumberOfExchanges = SeaPearl.IntVar(-n, 0, "minusNumberOfExchanges", trailer) 
+    SeaPearl.addVariable!(model, minusNumberOfExchanges; branchable=false)
     vars = SeaPearl.AbstractIntVar[]
+
+    #Concatenate all values of xNotEqualIndex_Int and minusNumberOfExchanges
     for i in 1:n
-        vars = cat(vars, int2bool(x_offset[i]); dims=1) #TODO int2bool?
+        vars = cat(vars, xNotEqualIndex_Int[i]; dims=1) 
```

Comment:
Just to be sure, we are trying to maximize the number of exchanges (ie. the number of times a peer is not linked to itself, is that right ?

Reply:
Yes, knowing that a peer linked to itself means that the peer does not participate in any cycle (in other words, the patient does not receive a kidney)